projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js

```diff
@@ -7,219 +7,272 @@ import { useSelect } from '@wordpress/data';
+	const [ isLoadingCategories, setIsLoadingCategories ] = useState( false );
+	const [ showAnimation, setShowAnimation ] = useState( false );
+	const [ needsMoreCharacters, setNeedsMoreCharacters ] = useState( 0 );
+	const [ showRetry, setShowRetry ] = useState( false );
+	const [ errorMessage, setErrorMessage ] = useState( false );
 
 	// Let's grab post data so that we can do something smart.
-	const currentPost = select( 'core/editor' ).getCurrentPost();
-	if ( ! currentPost ) {
-		return '';
-	}
+	const currentPostTitle = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'title' ),
+		[]
+	);
 
-	// If there is no title, there is not much we can do.
-	if ( ! currentPost.title ) {
-		return createPrompt( '', editorContent, '' );
-	}
+	let loading = false;
+	const categories = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'categories' ),
+		[]
+	);
 
-	// We are filtering out the default "Uncategorized"
-	const categories = currentPost.categories.filter( catId => catId !== 1 );
-	const tags = currentPost.tags;
+	const categoryObjects = useSelect(
+		select => {
+			return categories
+				.map( categoryId => {
+					const category = select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId );
 
-	// User did not set any categories, we are going to base the suggestions off a title.
-	if ( ! categories.length && ! tags.length ) {
-		return createPrompt( currentPost.title, editorContent, '' );
-	}
+					if ( ! category ) {
+						// Data is not yet loaded
+						loading = true;
+						return;
+					}
 
-	// We are grabbing more data from WP.
-	const categoryObjects = categories.map( categoryId =>
-		select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId )
+					return category;
+				} )
+				.filter( Boolean ); // Remove undefined values
+		},
+		[ categories ]
 	);
-	const tagObjects = tags.map( tagId =>
-		select( 'core' ).getEntityRecord( 'taxonomy', 'post_tag', tagId )
+
+	const tags = useSelect( select => select( 'core/editor' ).getEditedPostAttribute( 'tags' ), [] );
+
+	const tagObjects = useSelect(
+		select => {
+			return tags
+				.map( tagId => {
+					const tag = select( 'core' ).getEntityRecord( 'taxonomy', 'post_tag', tagId );
+
+					if ( ! tag ) {
+						// Data is not yet loaded
+						loading = true;
+						return;
+					}
+
+					return tag;
+				} )
+				.filter( Boolean ); // Remove undefined values
+		},
+		[ tags ]
 	);
 
-	// For now categories are combined with tags into the same object. I am not sure they should
-	const taxonomies = categoryObjects.concat( tagObjects );
+	useEffect( () => {
+		setIsLoadingCategories( loading );
+	}, [ loading ] );
 
-	// We want to wait until all category names are loaded. This will return empty string (aka loading state) until all objects are truthy.
-	if ( taxonomies.filter( obj => ! obj || ! obj.name ).length ) {
-		return '';
-	}
+	const createPrompt = () => {
```

Comment:
Having this in closure here is more spaghetti-like than the previous version

Reply:
I do not think so. Previous version was proper Italian-made with the finest water.

Reply:
Sure, sure, but you are using `contentBefore` in closure (line 108) before it is even declared (line 163).
Hoisting will make it work, but it is really confusing.